New module: azure_rm_expressroute

praveenghuge opened the pull request:
**SUMMARY**
This module introduces a wide variety of configurations for creating an express route on Azure.

> Create Express Route on Azure.
> Update Express Route on Azure.
> Delete Express Route on Azure.
> Get Express Route on Azure.
> List Express Route in Azure Resource Group.

Fixes #475

**ISSUE TYPE**

> New Module Pull Request


**COMPONENT NAME**
> azure_rm_expressroute

**Additional Support:**
for express route peering support we need to create resource in ansible https://docs.microsoft.com/en-us/azure/expressroute/expressroute-introduction
created issue https://github.com/ansible-collections/azure/issues/483

**ADDITIONAL INFORMATION**
This module allows user to create/update or delete the express route. It also provides all the parameters given by https://docs.microsoft.com/en-us/azure/expressroute/expressroute-introduction
More information: 

```- name: Create Express route (check mode)
  azure_rm_expressroute:
    location: eastus
    name: expressroute123
    resource_group: resourcegroup12
    allow_classic_operations: true
    global_reach_enabled: false
    tags:
      a: b
    authorizations:
      - name: authorization_test
    service_provider_properties:
      service_provider_name: Aryaka Networks
      peering_location: Seattle
      bandwidth_in_mbps: '200'
    sku:
      tier: premium
      family: metereddata```

Commit:
first commit

Commit:
fix file

Commit:
new changes

Commit:
working create/update/delete/get

pravin94ghuge committed:
change username

Fred-sun commented:
@praveenghuge Thank for your contribute! I will review and push for merge!

Commit:
minor changes

ikarldasan committed:
Fix sanity issues

ikarldasan committed:
Improve integration test

Commit:
remove return

Commit:
remove return

Fred-sun commented on plugins/modules/azure_rm_expressroute.py:
```diff
@@ -0,0 +1,361 @@
+DOCUMENTATION = '''
+---
+module: azure_rm_expressroute
+version_added: "0.1.2"
```
```suggestion
version_added: "1.5.1"
```

praveenghuge committed:
Merge branch 'dev' into expressroute

praveenghuge commented:
@Fred-sun can you please review

praveenghuge committed:
Merge branch 'dev' into expressroute

praveenghuge committed:
Merge branch 'dev' into expressroute

Fred-sun reviewed (commented):
> @Fred-sun can you please review

@praveenghuge Thanks for your contribution! I working in this!

pravin94ghuge committed:
address changes

Fred-sun commented on plugins/modules/azure_rm_expressroute.py:
```diff
@@ -0,0 +1,361 @@
+                description: Name of the authorization.
+                required: true
+            type: str
+        type: list
```
```suggestion

```

Fred-sun commented on plugins/modules/azure_rm_expressroute.py:
```diff
@@ -0,0 +1,361 @@
+        type: bool
+    authorizations:
+        description:
+            - The list of authorizations.
```
```suggestion
            - The list of authorizations.
        type: list
```

pravin94ghuge committed:
changes addressed

Fred-sun commented on plugins/modules/azure_rm_expressroute.py:
```diff
@@ -0,0 +1,361 @@
+            type: str
+        type: list
+    allow_classic_operations:
+        description: Support for classic operations.
```
```suggestion
        description:
            - Support for classic operations.
```

Fred-sun commented on plugins/modules/azure_rm_expressroute.py:
```diff
@@ -0,0 +1,361 @@
+          - absent
+          - present
+extends_documentation_fragment:
+- azure.azcollection.azure
```
```suggestion
    - azure.azcollection.azure
    - azure.azcollection.azure_tags
```

Fred-sun commented on plugins/modules/azure_rm_expressroute.py:
```diff
@@ -0,0 +1,361 @@
+'''
+
+RETURN = '''
+sample: {
```
@praveenghuge The return value should be describe as following doc:
```e.g.
state:
    description:
        - *****.
    returned: success
    type: complex
    contains:
        ****:
            description:
                - ****.
            returned: always
            type: str
            sample: "****"
......

Fred-sun commented on plugins/modules/azure_rm_expressroute_info.py:
```diff
@@ -0,0 +1,166 @@
+'''
+
+RETURN = '''
+
```
Like the following describe! thanks!
```e.g.
state:
    description:
        - *****.
    returned: success
    type: complex
    contains:
        ****:
            description:
                - ****.
            returned: always
            type: str
            sample: "****"

Fred-sun commented on plugins/modules/azure_rm_expressroute.py:
```diff
@@ -0,0 +1,361 @@
+    allow_classic_operations: true
+    global_reach_enabled: false
+    tags:
+       - a: b
```
```suggestion
      - a: b
```
```suggestion
       a: b
```

Fred-sun commented on plugins/modules/azure_rm_expressroute.py:
```diff
@@ -0,0 +1,361 @@
+    tags:
+       - a: b
+    authorizations:
+       - name: authorization_test
```
```suggestion
      - name: authorization_test
```

Fred-sun commented on tests/integration/targets/azure_rm_expressroute/tasks/main.yml:
```diff
@@ -0,0 +1,119 @@
+- name: Create random express route
+  set_fact:
+      express_route: "test{{ resource_group | hash('md5') | truncate(16, True, '') + (65535 | random | string) }}"
```
```suggestion
    express_route: "test{{ resource_group | hash('md5') | truncate(16, True, '') + (65535 | random | string) }}"
```

Fred-sun commented on plugins/modules/azure_rm_expressroute_info.py:
```diff
@@ -0,0 +1,166 @@
+
+
+extends_documentation_fragment:
+- azure.azcollection.azure
```
```suggestion
    - azure.azcollection.azure
```

Fred-sun commented on tests/integration/targets/azure_rm_expressroute/tasks/main.yml:
```diff
@@ -0,0 +1,119 @@
+# azure_rm_expressroute cleanup
+#
+- name: Delete Express Route
+    azure_rm_expressroute:
```
This block is not written to the YAML specification! Thank you very much!

Fred-sun commented on tests/integration/targets/azure_rm_expressroute/aliases:
```diff
@@ -0,0 +1,5 @@
+cloud/azure
+shippable/azure/group2
```
```suggestion
shippable/azure/group10
```

Fred-sun reviewed (commented):
@praveenghuge Would you please share the result of you execute the playbook "tests/integration/targets/azure_rm_expressroute/tasks/main.yml"? It will helpful to push merge this PR! Thank you very much!

praveenghuge commented:
playbook

```
---
 - hosts: localhost
   connection: local
   tasks:

    - name: Create Express route (check mode)
      azure.azcollection.azure_rm_expressroute:

        subscription_id: XXXXX
        tenant: XXXXX
        client_id: XXXX
        secret: XXXX
        resource_group: test
        name: test2
        location: eastus
        state: present
        allow_classic_operations: true
        global_reach_enabled: false
        tags:
                a: b
        authorizations:
                - name: authorization_test
        service_provider_properties:
                service_provider_name: Aryaka Networks
                peering_location: Seattle
                bandwidth_in_mbps: '200'
        sku:
                tier: premium
                family: metereddata
      check_mode: yes
```

output 
```
changed: [localhost] => {
    "changed": true,
    "check_mode": true,
    "invocation": {
        "module_args": {
            "ad_user": null,
            "adfs_authority_url": null,
            "allow_classic_operations": true,
            "api_profile": "latest",
            "append_tags": true,
            "auth_source": "auto",
            "authorizations": [
                {
                    "name": "authorization_test"
                }
            ],
            "cert_validation_mode": null,
            "client_id": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "cloud_environment": "AzureCloud",
            "global_reach_enabled": false,
            "location": "eastus",
            "name": "test2",
            "password": null,
            "profile": null,
            "resource_group": "test-praveen",
            "secret": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "service_provider_properties": {
                "bandwidth_in_mbps": "200",
                "peering_location": "Seattle",
                "service_provider_name": "Aryaka Networks"
            },
            "sku": {
                "family": "metereddata",
                "tier": "premium"
            },
            "state": "present",
            "subscription_id": "64caccf3-b508-41e7-92ed-d7ed95b32621",
            "tags": {
                "a": "b"
            },
            "tenant": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER"
        }
    },
    "state": {}
}
META: ran handlers
META: ran handlers

PLAY RECAP ***************************************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

praveenghuge commented:
playbook

```
---
 - hosts: localhost
   connection: local
   tasks:

    - name: Create Express route
      azure.azcollection.azure_rm_expressroute:

        subscription_id: XXXXX
        tenant: XXXXX
        client_id: XXXX
        secret: XXXX
        resource_group: test
        name: test2
        location: eastus
        state: present
        allow_classic_operations: true
        global_reach_enabled: false
        tags:
                a: b
        authorizations:
                - name: authorization_test
        service_provider_properties:
                service_provider_name: Aryaka Networks
                peering_location: Seattle
                bandwidth_in_mbps: '200'
        sku:
                tier: premium
                family: metereddata
```

output

```
changed: [localhost] => {
    "changed": true,
    "check_mode": false,
    "invocation": {
        "module_args": {
            "ad_user": null,
            "adfs_authority_url": null,
            "allow_classic_operations": true,
            "api_profile": "latest",
            "append_tags": true,
            "auth_source": "auto",
            "authorizations": [
                {
                    "name": "authorization_test"
                }
            ],
            "cert_validation_mode": null,
            "client_id": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "cloud_environment": "AzureCloud",
            "global_reach_enabled": false,
            "location": "eastus",
            "name": "test2",
            "password": null,
            "profile": null,
            "resource_group": "test-praveen",
            "secret": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "service_provider_properties": {
                "bandwidth_in_mbps": "200",
                "peering_location": "Seattle",
                "service_provider_name": "Aryaka Networks"
            },
            "sku": {
                "family": "metereddata",
                "name": "premium_metereddata",
                "tier": "premium"
            },
            "state": "present",
            "subscription_id": "64caccf3-b508-41e7-92ed-d7ed95b32621",
            "tags": {
                "a": "b"
            },
            "tenant": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER"
        }
    },
    "state": {
        "additional_properties": {},
        "allow_classic_operations": true,
        "authorizations": [
            {
                "authorization_key": "bbdb6919-ac1a-43a7-ab39-8bfa955ef23f",
                "authorization_use_status": "Available",
                "etag": "W/\"4a3ff51b-12cc-4972-9313-24421edbc855\"",
                "id": "/subscriptions/64caccf3-b508-41e7-92ed-d7ed95b32621/resourceGroups/test-praveen/providers/Microsoft.Network/expressRouteCircuits/test2/authorizations/authorization_test",
                "name": "authorization_test",
                "provisioning_state": "Succeeded",
                "type": "Microsoft.Network/expressRouteCircuits/authorizations"
            }
        ],
        "bandwidth_in_gbps": null,
        "circuit_provisioning_state": "Enabled",
        "express_route_port": null,
        "gateway_manager_etag": "",
        "global_reach_enabled": false,
        "id": "/subscriptions/64caccf3-b508-41e7-92ed-d7ed95b32621/resourceGroups/test-praveen/providers/Microsoft.Network/expressRouteCircuits/test2",
        "location": "eastus",
        "name": "test2",
        "peerings": [],
        "provisioning_state": "Succeeded",
        "service_key": "b9e910e0-8475-43c0-b16b-f317fc6ad427",
        "service_provider_notes": null,
        "service_provider_provisioning_state": "NotProvisioned",
        "stag": 24,
        "tags": {
            "a": "b"
        },
        "type": "Microsoft.Network/expressRouteCircuits"
    }
}
META: ran handlers
META: ran handlers

PLAY RECAP ***************************************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

praveenghuge commented:
playbook

```
---
 - hosts: localhost
   connection: local
   tasks:

    - name: Create Express route (idempotent)
      azure.azcollection.azure_rm_expressroute:

        subscription_id: XXXXX
        tenant: XXXXX
        client_id: XXXX
        secret: XXXX
        resource_group: test
        name: test2
        location: eastus
        state: present
        allow_classic_operations: true
        global_reach_enabled: false
        tags:
                a: b
        authorizations:
                - name: authorization_test
        service_provider_properties:
                service_provider_name: Aryaka Networks
                peering_location: Seattle
                bandwidth_in_mbps: '200'
        sku:
                tier: premium
                family: metereddata
```

output
```
ok: [localhost] => {
    "changed": false,
    "check_mode": false,
    "invocation": {
        "module_args": {
            "ad_user": null,
            "adfs_authority_url": null,
            "allow_classic_operations": true,
            "api_profile": "latest",
            "append_tags": true,
            "auth_source": "auto",
            "authorizations": [
                {
                    "name": "authorization_test"
                }
            ],
            "cert_validation_mode": null,
            "client_id": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "cloud_environment": "AzureCloud",
            "global_reach_enabled": false,
            "location": "eastus",
            "name": "test2",
            "password": null,
            "profile": null,
            "resource_group": "test-praveen",
            "secret": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "service_provider_properties": {
                "bandwidth_in_mbps": "200",
                "peering_location": "Seattle",
                "service_provider_name": "Aryaka Networks"
            },
            "sku": {
                "family": "metereddata",
                "tier": "premium"
            },
            "state": "present",
            "subscription_id": "64caccf3-b508-41e7-92ed-d7ed95b32621",
            "tags": {
                "a": "b"
            },
            "tenant": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER"
        }
    },
    "state": {
        "additional_properties": {},
        "allow_classic_operations": true,
        "authorizations": [
            {
                "authorization_key": "bbdb6919-ac1a-43a7-ab39-8bfa955ef23f",
                "authorization_use_status": "Available",
                "etag": "W/\"4a3ff51b-12cc-4972-9313-24421edbc855\"",
                "id": "/subscriptions/64caccf3-b508-41e7-92ed-d7ed95b32621/resourceGroups/test-praveen/providers/Microsoft.Network/expressRouteCircuits/test2/authorizations/authorization_test",
                "name": "authorization_test",
                "provisioning_state": "Succeeded",
                "type": "Microsoft.Network/expressRouteCircuits/authorizations"
            }
        ],
        "bandwidth_in_gbps": null,
        "circuit_provisioning_state": "Enabled",
        "express_route_port": null,
        "gateway_manager_etag": "",
        "global_reach_enabled": false,
        "id": "/subscriptions/64caccf3-b508-41e7-92ed-d7ed95b32621/resourceGroups/test-praveen/providers/Microsoft.Network/expressRouteCircuits/test2",
        "location": "eastus",
        "name": "test2",
        "peerings": [],
        "provisioning_state": "Succeeded",
        "service_key": "b9e910e0-8475-43c0-b16b-f317fc6ad427",
        "service_provider_notes": null,
        "service_provider_provisioning_state": "NotProvisioned",
        "stag": 24,
        "tags": {
            "a": "b"
        },
        "type": "Microsoft.Network/expressRouteCircuits"
    }
}
META: ran handlers
META: ran handlers

PLAY RECAP ***************************************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

praveenghuge commented:
playbook

```
---
 - hosts: localhost
   connection: local
   tasks:

    - name: Update Express route
      azure.azcollection.azure_rm_expressroute:

        subscription_id: XXXXX
        tenant: XXXXX
        client_id: XXXXX
        secret: XXXX
        resource_group: test-praveen
        name: test2
        location: eastus
        state: present
        allow_classic_operations: true
        global_reach_enabled: false
        tags:
                a: b
                test: modified
        authorizations:
                - name: authorization_test
        service_provider_properties:
                service_provider_name: Aryaka Networks
                peering_location: Seattle
                bandwidth_in_mbps: '200'
        sku:
                tier: premium
                family: metereddata
```

output

```
changed: [localhost] => {
    "changed": true,
    "check_mode": false,
    "invocation": {
        "module_args": {
            "ad_user": null,
            "adfs_authority_url": null,
            "allow_classic_operations": true,
            "api_profile": "latest",
            "append_tags": true,
            "auth_source": "auto",
            "authorizations": [
                {
                    "name": "authorization_test"
                }
            ],
            "cert_validation_mode": null,
            "client_id": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "cloud_environment": "AzureCloud",
            "global_reach_enabled": false,
            "location": "eastus",
            "name": "test2",
            "password": null,
            "profile": null,
            "resource_group": "test-praveen",
            "secret": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "service_provider_properties": {
                "bandwidth_in_mbps": "200",
                "peering_location": "Seattle",
                "service_provider_name": "Aryaka Networks"
            },
            "sku": {
                "family": "metereddata",
                "name": "premium_metereddata",
                "tier": "premium"
            },
            "state": "present",
            "subscription_id": "64caccf3-b508-41e7-92ed-d7ed95b32621",
            "tags": {
                "a": "b",
                "test": "modified"
            },
            "tenant": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER"
        }
    },
    "state": {
        "additional_properties": {},
        "allow_classic_operations": true,
        "authorizations": [
            {
                "authorization_key": "bbdb6919-ac1a-43a7-ab39-8bfa955ef23f",
                "authorization_use_status": "Available",
                "etag": "W/\"4590b630-dc31-4885-abb8-551de8c0466e\"",
                "id": "/subscriptions/64caccf3-b508-41e7-92ed-d7ed95b32621/resourceGroups/test-praveen/providers/Microsoft.Network/expressRouteCircuits/test2/authorizations/authorization_test",
                "name": "authorization_test",
                "provisioning_state": "Succeeded",
                "type": "Microsoft.Network/expressRouteCircuits/authorizations"
            }
        ],
        "bandwidth_in_gbps": null,
        "circuit_provisioning_state": "Enabled",
        "express_route_port": null,
        "gateway_manager_etag": "",
        "global_reach_enabled": false,
        "id": "/subscriptions/64caccf3-b508-41e7-92ed-d7ed95b32621/resourceGroups/test-praveen/providers/Microsoft.Network/expressRouteCircuits/test2",
        "location": "eastus",
        "name": "test2",
        "peerings": [],
        "provisioning_state": "Succeeded",
        "service_key": "b9e910e0-8475-43c0-b16b-f317fc6ad427",
        "service_provider_notes": null,
        "service_provider_provisioning_state": "NotProvisioned",
        "stag": 24,
        "tags": {
            "a": "b",
            "test": "modified"
        },
        "type": "Microsoft.Network/expressRouteCircuits"
    }
}
META: ran handlers
META: ran handlers

PLAY RECAP ***************************************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

praveenghuge commented:
playbook

```
---
 - hosts: localhost
   connection: local
   tasks:

    - name: Delete Express route
      azure.azcollection.azure_rm_expressroute:
        subscription_id: XXXX
        tenant: XXXX
        client_id: XXXX
        secret: XXXX
        resource_group: test-praveen
        name: test2
        state: absent
```

output

```
changed: [localhost] => {
    "changed": true,
    "check_mode": false,
    "invocation": {
        "module_args": {
            "ad_user": null,
            "adfs_authority_url": null,
            "allow_classic_operations": null,
            "api_profile": "latest",
            "append_tags": true,
            "auth_source": "auto",
            "authorizations": null,
            "cert_validation_mode": null,
            "client_id": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "cloud_environment": "AzureCloud",
            "global_reach_enabled": null,
            "location": null,
            "name": "test2",
            "password": null,
            "profile": null,
            "resource_group": "test-praveen",
            "secret": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "service_provider_properties": null,
            "sku": null,
            "state": "absent",
            "subscription_id": "64caccf3-b508-41e7-92ed-d7ed95b32621",
            "tags": null,
            "tenant": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER"
        }
    },
    "state": {
        "additional_properties": {},
        "allow_classic_operations": true,
        "authorizations": [
            {
                "authorization_key": "bbdb6919-ac1a-43a7-ab39-8bfa955ef23f",
                "authorization_use_status": "Available",
                "etag": "W/\"4590b630-dc31-4885-abb8-551de8c0466e\"",
                "id": "/subscriptions/64caccf3-b508-41e7-92ed-d7ed95b32621/resourceGroups/test-praveen/providers/Microsoft.Network/expressRouteCircuits/test2/authorizations/authorization_test",
                "name": "authorization_test",
                "provisioning_state": "Succeeded",
                "type": "Microsoft.Network/expressRouteCircuits/authorizations"
            }
        ],
        "bandwidth_in_gbps": null,
        "circuit_provisioning_state": "Enabled",
        "express_route_port": null,
        "gateway_manager_etag": "",
        "global_reach_enabled": false,
        "id": "/subscriptions/64caccf3-b508-41e7-92ed-d7ed95b32621/resourceGroups/test-praveen/providers/Microsoft.Network/expressRouteCircuits/test2",
        "location": "eastus",
        "name": "test2",
        "peerings": [],
        "provisioning_state": "Succeeded",
        "service_key": "b9e910e0-8475-43c0-b16b-f317fc6ad427",
        "service_provider_notes": null,
        "service_provider_provisioning_state": "NotProvisioned",
        "stag": 24,
        "status": "Deleted",
        "tags": {
            "a": "b",
            "test": "modified"
        },
        "type": "Microsoft.Network/expressRouteCircuits"
    }
}
META: ran handlers
META: ran handlers

PLAY RECAP ***************************************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

praveenghuge commented:
playbook

---
 - hosts: localhost
   connection: local
   tasks:

    - name: Delete Express route (idempotent)
      azure.azcollection.azure_rm_expressroute:
        subscription_id: XXXX
        tenant: XXXX
        client_id: XXXX
        secret: XXXX
        resource_group: test-praveen
        name: test2
        state: absent
```

output
```
ok: [localhost] => {
    "changed": false,
    "check_mode": false,
    "invocation": {
        "module_args": {
            "ad_user": null,
            "adfs_authority_url": null,
            "allow_classic_operations": null,
            "api_profile": "latest",
            "append_tags": true,
            "auth_source": "auto",
            "authorizations": null,
            "cert_validation_mode": null,
            "client_id": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "cloud_environment": "AzureCloud",
            "global_reach_enabled": null,
            "location": null,
            "name": "test2",
            "password": null,
            "profile": null,
            "resource_group": "test-praveen",
            "secret": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "service_provider_properties": null,
            "sku": null,
            "state": "absent",
            "subscription_id": "64caccf3-b508-41e7-92ed-d7ed95b32621",
            "tags": null,
            "tenant": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER"
        }
    },
    "state": {}
}
META: ran handlers
META: ran handlers

PLAY RECAP ***************************************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

praveenghuge commented:
playbook 

```
---
 - hosts: localhost
   connection: local
   tasks:

    - name: Get list of Express route
      azure.azcollection.azure_rm_expressroute_info:

        subscription_id: 64caccf3-b508-41e7-92ed-d7ed95b32621
        tenant: eff7f985-dc58-4935-a906-050609be85c3
        client_id: 4a5c3ed8-83d3-4bda-9f31-39129ab2f4b6
        secret: /@SMLXK@dc.5bxY0CR1t0ed[d@EhQW5[
        resource_group: test-praveen
```

output

```
ok: [localhost] => {
    "changed": false,
    "expressroute": [
        {
            "additional_properties": {},
            "allow_classic_operations": true,
            "authorizations": [
                {
                    "authorization_key": "af538d56-d627-42bb-9ca1-fd37a3a33d05",
                    "authorization_use_status": "Available",
                    "etag": "W/\"b39a3fa0-4c21-42af-a59c-82886ba8d922\"",
                    "id": "/subscriptions/64caccf3-b508-41e7-92ed-d7ed95b32621/resourceGroups/test-praveen/providers/Microsoft.Network/expressRouteCircuits/test2/authorizations/authorization_test",
                    "name": "authorization_test",
                    "provisioning_state": "Succeeded",
                    "type": "Microsoft.Network/expressRouteCircuits/authorizations"
                }
            ],
            "bandwidth_in_gbps": null,
            "circuit_provisioning_state": "Enabled",
            "express_route_port": null,
            "gateway_manager_etag": "",
            "global_reach_enabled": false,
            "id": "/subscriptions/64caccf3-b508-41e7-92ed-d7ed95b32621/resourceGroups/test-praveen/providers/Microsoft.Network/expressRouteCircuits/test2",
            "location": "eastus",
            "name": "test2",
            "peerings": [],
            "provisioning_state": "Succeeded",
            "service_key": "4b9caeb9-dc34-4bab-acd2-a10ed3212e13",
            "service_provider_notes": null,
            "service_provider_provisioning_state": "NotProvisioned",
            "stag": 24,
            "tags": {
                "a": "b",
                "test": "modified"
            },
            "type": "Microsoft.Network/expressRouteCircuits"
        }
    ],
    "invocation": {
        "module_args": {
            "ad_user": null,
            "adfs_authority_url": null,
            "api_profile": "latest",
            "auth_source": "auto",
            "cert_validation_mode": null,
            "client_id": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "cloud_environment": "AzureCloud",
            "name": null,
            "password": null,
            "profile": null,
            "resource_group": "test-praveen",
            "secret": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "subscription_id": "64caccf3-b508-41e7-92ed-d7ed95b32621",
            "tenant": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER"
        }
    }
}
META: ran handlers
META: ran handlers

PLAY RECAP ***************************************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

praveenghuge commented:
playbook

```
---
 - hosts: localhost
   connection: local
   tasks:

    - name: Get Express route
      azure.azcollection.azure_rm_expressroute_info:

        subscription_id: XXXX
        tenant: XXXX
        client_id: XXX
        secret: XXX
        resource_group: test-praveen
        name: test2
```

output

```
ok: [localhost] => {
    "changed": false,
    "expressroute": [
        {
            "additional_properties": {},
            "allow_classic_operations": true,
            "authorizations": [
                {
                    "authorization_key": "af538d56-d627-42bb-9ca1-fd37a3a33d05",
                    "authorization_use_status": "Available",
                    "etag": "W/\"b39a3fa0-4c21-42af-a59c-82886ba8d922\"",
                    "id": "/subscriptions/Xxxx/resourceGroups/test-praveen/providers/Microsoft.Network/expressRouteCircuits/test2/authorizations/authorization_test",
                    "name": "authorization_test",
                    "provisioning_state": "Succeeded",
                    "type": "Microsoft.Network/expressRouteCircuits/authorizations"
                }
            ],
            "bandwidth_in_gbps": null,
            "circuit_provisioning_state": "Enabled",
            "express_route_port": null,
            "gateway_manager_etag": "",
            "global_reach_enabled": false,
            "id": "/subscriptions/XXXX/resourceGroups/test-praveen/providers/Microsoft.Network/expressRouteCircuits/test2",
            "location": "eastus",
            "name": "test2",
            "peerings": [],
            "provisioning_state": "Succeeded",
            "service_key": "4b9caeb9-dc34-4bab-acd2-a10ed3212e13",
            "service_provider_notes": null,
            "service_provider_provisioning_state": "NotProvisioned",
            "stag": 24,
            "tags": {
                "a": "b",
                "test": "modified"
            },
            "type": "Microsoft.Network/expressRouteCircuits"
        }
    ],
    "invocation": {
        "module_args": {
            "ad_user": null,
            "adfs_authority_url": null,
            "api_profile": "latest",
            "auth_source": "auto",
            "cert_validation_mode": null,
            "client_id": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "cloud_environment": "AzureCloud",
            "name": "test2",
            "password": null,
            "profile": null,
            "resource_group": "test-praveen",
            "secret": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "subscription_id": "xxxx",
            "tenant": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER"
        }
    }
}
META: ran handlers
META: ran handlers

PLAY RECAP ***************************************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

pravin94ghuge committed:
address changes

pravin94ghuge committed:
address changes

pravin94ghuge committed:
small change

praveenghuge commented:
@Fred-sun addressed the changes requested.
Thanks for the review 👍

praveenghuge committed:
Merge branch 'dev' into expressroute

praveenghuge committed:
Merge branch 'dev' into expressroute

praveenghuge committed:
Merge branch 'dev' into expressroute

haiyuazhang committed:
fixing update account_enabled bug in azure_rm_aduser.py (#524)

haiyuazhang committed:
fixing ad related auth issue when using service pricinpal. (#525)

haiyuazhang committed:
change class name of azure_rm_aduser (#526)

* class are worngly named. fixed.

* fixing sanity errors.

praveenghuge committed:
Merge branch 'dev' into expressroute

Fred-sun commented on plugins/modules/azure_rm_expressroute.py:
```diff
@@ -0,0 +1,380 @@
+        }
+'''
+
+from __future__ import absolute_import, division, print_function
```
from __future__ imports must occur at the beginning of the file

Fred-sun commented on plugins/modules/azure_rm_expressroute.py:
```diff
@@ -0,0 +1,380 @@
+'''
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
```
The '__metaclass__" must occur at the beginning of the file

Fred-sun commented on plugins/modules/azure_rm_expressroute.py:
```diff
@@ -0,0 +1,380 @@
+                required: true
+            type: str
+    allow_classic_operations:
+        description: 
```
trailing whitespace
```suggestion
        description:
```

Fred-sun commented on plugins/modules/azure_rm_expressroute.py:
```diff
@@ -0,0 +1,380 @@
+        description:
+            - The name of the SKU.
+            - Please see L(https://azure.microsoft.com/en-in/pricing/details/expressroute/)
+        required: if state is present
```
```suggestion
            - Required sku when I(state=present).
```

Fred-sun commented on plugins/modules/azure_rm_expressroute.py:
```diff
@@ -0,0 +1,380 @@
+            name:
+                description: Name of the authorization.
+                required: true
+            type: str
```
```suggestion
                type: str
```

praveenghuge committed:
Merge branch 'dev' into expressroute

pravin94ghuge committed:
address cahnges

praveenghuge commented:
@Fred-sun addressed the changes requested.
Thanks for the review 👍

praveenghuge committed:
Merge branch 'dev' into expressroute

praveenghuge committed:
Merge branch 'dev' into expressroute

praveenghuge committed:
Merge branch 'dev' into expressroute

praveenghuge commented:
Hi @Fred-sun can you please review the PR. addressed the changes
thank you so much 👍

Fred-sun commented on plugins/modules/azure_rm_expressroute.py:
```diff
@@ -0,0 +1,379 @@
+    returned: always
+    type: dict
+
+    sample: {
```
@praveenghuge If you seperate this sample（same with azure_rm_expressroute_info.py), it will be even more perfect! Thank you very much!
```e.g.
return_paramter:
    description:
        - *******
    returned: *
    type: *
    sample: *

Fred-sun commented on tests/sanity/ignore-2.11.txt:
```diff
@@ -354,5 +354,13 @@ plugins/modules/azure_rm_openshiftmanagedcluster.py validate-modules:parameter-l
 plugins/modules/azure_rm_subnet.py validate-modules:doc-elements-mismatch
 plugins/modules/azure_rm_vmbackuppolicy.py validate-modules:parameter-list-no-elements
 plugins/modules/azure_rm_privatednsrecordset.py validate-modules:invalid-ansiblemodule-schema
+plugins/modules/azure_rm_privatednsrecordset.py validate-modules:required_if-requirements-unknown
```
@praveenghuge Please help to delete these ignores on ignore2.11.txt, 2.9.txt, 2.10.txt and 2.12.txt. It has been fixed by #540!

Fred-sun commented on plugins/modules/azure_rm_expressroute.py:
```diff
@@ -0,0 +1,379 @@
+            - Support for classic operations.
+        type: bool
+    state:
+      description:
```
Move back two Spaces（line 87~94)!

Fred-sun commented on plugins/modules/azure_rm_expressroute.py:
```diff
@@ -0,0 +1,379 @@
+            tier:
+                description:
+                    - The tier of the SKU
+                required: true
```
```suggestion
                type: str
                required: true
```

Fred-sun commented on plugins/modules/azure_rm_expressroute.py:
```diff
@@ -0,0 +1,379 @@
+            family:
+                description:
+                    - the family of the SKU
+                required: true
```
```suggestion
                type: str
                required: true
```

Fred-sun commented on plugins/modules/azure_rm_expressroute.py:
```diff
@@ -0,0 +1,379 @@
+        description:
+            - The name of the SKU.
+            - Please see L(https://azure.microsoft.com/en-in/pricing/details/expressroute/)
+            - Required sku when I(state=present).
```
```suggestion
            - Required sku when I(state=present).
        type: dict
```

Fred-sun commented on plugins/modules/azure_rm_expressroute.py:
```diff
@@ -0,0 +1,379 @@
+                    - UnlimitedData
+    global_reach_enabled:
+        description: Flag denoting global reach status.
+        default: true
```
Argument 'global_reach_enabled' in argument_spec defines default as (None) but documentation defines default as (True)

```suggestion

```

Fred-sun commented on plugins/modules/azure_rm_expressroute.py:
```diff
@@ -0,0 +1,379 @@
+                description:
+                    - The tier of the SKU
+                required: true
+                default: Standard
```
Argument 'tier' in argument_spec found in sku defines default as (None) but documentation defines default as ('Standard')

```suggestion

```

Fred-sun commented on plugins/modules/azure_rm_expressroute.py:
```diff
@@ -0,0 +1,379 @@
+    authorizations:
+        description:
+            - The list of authorizations.
+        type: list
```
```suggestion
        type: list
        elements: dict
```

Fred-sun commented on plugins/modules/azure_rm_expressroute.py:
```diff
@@ -0,0 +1,379 @@
+DOCUMENTATION = '''
+---
+module: azure_rm_expressroute
+version_added: "1.5.1"
```
```suggestion
version_added: "1.6.0"
```

Fred-sun commented on plugins/modules/azure_rm_expressroute_info.py:
```diff
@@ -0,0 +1,202 @@
+
+DOCUMENTATION = '''
+---
+module: azure_rm_expressroute_info
```
```suggestion
module: azure_rm_expressroute_info
version_added: "1.6.0"
```

Fred-sun commented on tests/integration/targets/azure_rm_expressroute/tasks/main.yml:
```diff
@@ -0,0 +1,119 @@
+- name: Test idempotent
+  azure_rm_expressroute:
+    name: "{{ express_route }}"
+    resource_group: {{ resource_group }}
```
```suggestion
    resource_group: "{{ resource_group }}"
```

Fred-sun commented on tests/integration/targets/azure_rm_expressroute/tasks/main.yml:
```diff
@@ -0,0 +1,119 @@
+- name: Update Express route
+  azure_rm_expressroute:
+    location: eastus
+    name: {{ express_route }}
```
```suggestion
    name: "{{ express_route }}"
```

Fred-sun commented on tests/integration/targets/azure_rm_expressroute/tasks/main.yml:
```diff
@@ -0,0 +1,119 @@
+  azure_rm_expressroute:
+    location: eastus
+    name: {{ express_route }}
+    resource_group: {{ resource_group }}
```
```suggestion
    resource_group: "{{ resource_group }}"
```

Fred-sun commented on tests/integration/targets/azure_rm_expressroute/tasks/main.yml:
```diff
@@ -0,0 +1,119 @@
+  assert:
+    that: 
+      - results.expressroute[0].tags.test == 'modified'
+      - zones.expressroute[0].type == 'Microsoft.Network/expressRouteCircuits'
```
```suggestion
      - results.expressroute[0].type == 'Microsoft.Network/expressRouteCircuits'
```

pravin94ghuge committed:
new changes

pravin94ghuge committed:
new changes

Fred-sun commented on plugins/modules/azure_rm_expressroute.py:
```diff
@@ -0,0 +1,375 @@
+        self.results['check_mode'] = self.check_mode
+
+        # retrieve resource group to make sure it exists
+        self.get_resource_group(self.resource_group)
```
```suggestion
        resource_group = self.get_resource_group(self.resource_group)
        if not self.location:
            # Set default location
            self.module.params['location'] = resource_group.location
```

Fred-sun commented on plugins/modules/azure_rm_expressroute.py:
```diff
@@ -0,0 +1,375 @@
+
+        super(AzureExpressRoute, self).__init__(self.module_arg_spec,
+                                                supports_check_mode=True,
+                                                supports_tags=True,
```
```suggestion
                                                supports_tags=True)
```

Fred-sun commented on plugins/modules/azure_rm_expressroute.py:
```diff
@@ -0,0 +1,375 @@
+        super(AzureExpressRoute, self).__init__(self.module_arg_spec,
+                                                supports_check_mode=True,
+                                                supports_tags=True,
+                                                required_if=required_if)
```
```suggestion

```

Fred-sun commented on plugins/modules/azure_rm_expressroute.py:
```diff
@@ -0,0 +1,375 @@
+            global_reach_enabled=dict(type='bool'),
+        )
+
+        required_if = [
```
Please delete line 221~224, If we delete the resource, It just need resource_group , name and state paramter! Thank you very much!
```suggestion

```

Fred-sun commented on plugins/modules/azure_rm_expressroute.py:
```diff
@@ -0,0 +1,375 @@
+        self.allow_classic_operations = None
+        self.authorizations = None
+        self.service_provider_properties = None
+        self. global_reach_enabled = None
```
```suggestion
        self.global_reach_enabled = None
```

pravin94ghuge committed:
addrss changes

pravin94ghuge committed:
address changes

pravin94ghuge committed:
address changes

praveenghuge commented:
Hi @Fred-sun can you please review the PR. addressed the changes
thank you so much 👍

Fred-sun commented:
> Hi @Fred-sun can you please review the PR. addressed the changes
> thank you so much 👍

I'm working on it!

Fred-sun commented on plugins/modules/azure_rm_expressroute_info.py:
```diff
@@ -0,0 +1,203 @@
+
+extends_documentation_fragment:
+    - azure.azcollection.azure
+    - azure.azcollection.azure_tags
```
```suggestion

```

Fred-sun commented on plugins/modules/azure_rm_expressroute.py:
```diff
@@ -0,0 +1,371 @@
+        type: str
+        default: present
+        choices:
+             - absent
```
```suggestion
            - absent
```

Fred-sun commented on plugins/modules/azure_rm_expressroute.py:
```diff
@@ -0,0 +1,371 @@
+                    - MeteredData
+                    - UnlimitedData
+    global_reach_enabled:
+        description: Flag denoting global reach status.
```
```suggestion
        description:
            - Flag denoting global reach status.
```

Fred-sun commented on plugins/modules/azure_rm_expressroute.py:
```diff
@@ -0,0 +1,371 @@
+                type: str
+                required: true
+                choices:
+                    - MeteredData
```
```suggestion
                    - metereddata
```

Fred-sun commented on plugins/modules/azure_rm_expressroute.py:
```diff
@@ -0,0 +1,371 @@
+                required: true
+                choices:
+                    - MeteredData
+                    - UnlimitedData
```
```suggestion
                    - unlimiteddata
```

Fred-sun commented on plugins/modules/azure_rm_expressroute.py:
```diff
@@ -0,0 +1,371 @@
+                    - The tier of the SKU
+                type: str
+                choices:
+                    - Standard
```
```suggestion
                    - standard
```

Fred-sun commented on plugins/modules/azure_rm_expressroute.py:
```diff
@@ -0,0 +1,371 @@
+                type: str
+                choices:
+                    - Standard
+                    - Premium
```
```suggestion
                    - premium
```

Fred-sun commented on plugins/modules/azure_rm_expressroute.py:
```diff
@@ -0,0 +1,371 @@
+            tier:
+                description:
+                    - The tier of the SKU
+                type: str
```
```suggestion
                type: str
                required: true
```

Commit:
address changes

Commit:
merge upstream changes

Fred-sun commented on plugins/modules/azure_rm_expressroute.py:
```diff
@@ -0,0 +1,372 @@
+                    - metereddata
+                    - unlimiteddata
+    global_reach_enabled:
+        description: 
```
```suggestion
        description:
```

Commit:
merge upstream changes

praveenghuge commented:
Hi @Fred-sun can you please review the PR. addressed the changes
thank you so much 👍

Fred-sun commented on plugins/modules/azure_rm_expressroute_info.py:
```diff
@@ -0,0 +1,203 @@
+        self.tags = None
+
+        super(AzureExpressRouteInfo, self).__init__(
+            self.module_arg_spec, supports_tags=False)
```
```suggestion
            self.module_arg_spec, supports_tags=True)
```

Fred-sun commented on plugins/modules/azure_rm_expressroute.py:
```diff
@@ -0,0 +1,372 @@
+            - absent
+            - present
+extends_documentation_fragment:
+    - azure.azcollection.azure
```
```suggestion
    - azure.azcollection.azure
    - azure.azcollection.azure_tags
```

Commit:
address changes

praveenghuge commented:
Hi @Fred-sun can you please review the PR. addressed the changes
thank you so much 👍

praveenghuge committed:
Merge branch 'dev' into expressroute

praveenghuge commented:
@Fred-sun anything pending from our side. Thanks you so much 👍

Fred-sun commented:
@praveenghuge It is not pending, the new version will be released temporarily, and the PR will be merged after the new version is released. Thank you very much!

pravin94ghuge committed:
update version

praveenghuge committed:
Merge branch 'dev' into expressroute

praveenghuge commented:
@Fred-sun updated the latest version
Can you please review
Thank you so much 👍

Fred-sun commented:
@praveenghuge Greate! ready_for_review!

praveenghuge committed:
Merge branch 'dev' into expressroute

xuzhang3 commented:
LGTM
